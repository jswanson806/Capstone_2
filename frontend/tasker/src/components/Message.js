import React from "react";

const Message = ({message, preview, user, currUser, onConvoClick}) => {

    const handleClick = (user, currUserId, messageConvoId) => {
        onConvoClick(user, currUserId, messageConvoId);
    }

    const bodyClassName = message.sent_by === currUser.id ? "message-container-right" : "message-container-left";


    return (
        <div className="message-container">
            {!preview && (
            <div className={bodyClassName}>
                <p>{message.body}</p>
                <small className="message-timestamp">{message.created_at}</small>
            </div>)}

            {preview && (
                <div className="message-preview" 
                    onClick={() => handleClick(user, currUser.id, message.convo_id)}
                >
                    <h4>{user.firstName} {user.lastName.slice(0,1 + ".")}</h4>
                    <p>{message.body.slice(30 + '...')}</p>
                    <small className="message-preview-timestamp">{message.created_at}</small>
                </div>
            )}
        </div>
    )
};

export default Message;