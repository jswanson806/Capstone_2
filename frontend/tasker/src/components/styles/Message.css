.message-container {
    position: relative; /* Create a positioning context */
    padding: 10px;
    margin-bottom: 5px;
  }
  
  /* Style for the "New" badge */
  .message-container .badge {
    position: absolute;
    top: 5px;
    right: 5px;
  }
  
  .message-container .message-preview {
    display: flex;
    flex-direction: column;
  }
  
  .message-container .message-preview h4 {
    margin: 0;
  }
  
  .message-container .message-preview-text {
    margin-top: 10px;
    text-align: left; /* Align message text to the left */
  }
  
  .message-container .message-preview .message-preview-timestamp {
    margin-top: auto; /* Push the timestamp to the bottom */
    text-align: right; /* Align timestamp to the right */
  }

  .message-container-left {
    font-weight: bold;
    background-color: #B0E0E6;
    border-radius: 10px;
    padding: 10px;
    width: 70%;
    text-align: left;
  }

  .message-container-right {
    font-weight: bold;
    background-color: #FFB6C1;
    border-radius: 10px;
    padding: 10px;
    width: 70%;
    margin-left: 30%;
    text-align: right;
  }

  .modal-header {
    display: block !important;
    margin-top: -20px !important;
  }