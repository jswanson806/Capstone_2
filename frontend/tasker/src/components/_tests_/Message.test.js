import React from "react";
import {act, render, screen, fireEvent, waitFor} from '@testing-library/react';
import TaskerApi from '../../api.js';
import Message from "../Message.js";

const onConvoClick = jest.fn();
const currUser = {id: 1};
const user = {id: 2, email: "email@test.com", firstName: "Joe", lastName: "Doe"}
const message1 = {id: 1, body: "test message 1 to test the message", sent_by: 2, sent_to: 1, is_read: false, convo_id: 'u1u2j1'};
const message2 = {id: 1, body: "test message 2", is_read: true};

const updateMessage = jest.spyOn(TaskerApi, 'updateMessage');
const getSingleMessage = jest.spyOn(TaskerApi, 'getSingleMessage');

describe('smoke and snapshot testing', () => {
    test('renders without crashing', async () => {
        await act(async () => {
            render(
                <Message
                    onConvoClick={onConvoClick}
                    currUser={currUser}
                    user={user}
                    preview={false}
                    message={message1}
                />
            );
        });

    });

    test('matches snapshot', async () => {
        let asFragment;

        await act(async () => {
            const {asFragment: fragment} = render(
                <Message
                    onConvoClick={onConvoClick}
                    currUser={currUser}
                    user={user}
                    preview={false}
                    message={message1}
                />
            );
            asFragment = fragment;
        });

        expect(asFragment).toMatchSnapshot();
    });

});

describe('unread message displays badge', () => {

    test('works - displays on unread message', async () => {

        await act(async () => {
            render(
                <Message
                    onConvoClick={onConvoClick}
                    currUser={currUser}
                    user={user}
                    preview={false}
                    message={message1} // is_read = false
                />
            );
        });

        const badge = screen.queryByText("New");
        expect(badge).toBeInTheDocument();

    });

    test('works - does not display on read messages', async () => {


        await act(async () => {
            render(
                <Message
                    onConvoClick={onConvoClick}
                    currUser={currUser}
                    user={user}
                    preview={false}
                    message={message2} //is_read = true
                />
            );

        });

        const badge = screen.queryByText("New");
        expect(badge).not.toBeInTheDocument();

    });

});

describe('preview message tests', () => {

    test('works - renders preview message', async () => {
        await act(async () => {
            render(
                <Message
                    onConvoClick={onConvoClick}
                    currUser={currUser}
                    user={user}
                    preview={true}
                    message={message1} // is_read = false
                />
            );
        });

        const bodyPreview = screen.queryByTestId("message-preview");
        const bodyFull = screen.queryByText("message-body")
        
        expect(bodyPreview).toBeInTheDocument();
        expect(bodyFull).not.toBeInTheDocument();
    });

    test('message preview click calls api with correct data', async () => {

        updateMessage.mockResolvedValueOnce('response');
        getSingleMessage.mockResolvedValueOnce( {id: 1, body: "test message 1 to test the message", sent_by: 2, sent_to: 1, is_read: true} );

        await act(async () => {
            render(
                <Message
                    onConvoClick={onConvoClick}
                    currUser={currUser}
                    user={user}
                    preview={true}
                    message={message1} // is_read = false
                />
            );
        });

        // click the message preview
        const bodyPreview = screen.queryByTestId("message-preview");
        fireEvent.click(bodyPreview);

        // data to update message with
        const messageUpdate = {message: {is_read: true}};

        await waitFor(() => {
            // call onConvoClick
            expect(onConvoClick).toHaveBeenCalledTimes(1);
            expect(onConvoClick).toHaveBeenCalledWith(user, currUser.id, message1.convo_id);
            // call api updateMessage
            expect(updateMessage).toHaveBeenCalledTimes(1);
            expect(updateMessage).toHaveBeenCalledWith(1, messageUpdate);
            // call api getSingleMessage
            expect(getSingleMessage).toHaveBeenCalledTimes(1);
            expect(getSingleMessage).toHaveBeenCalledWith(1);
        });
    });

});