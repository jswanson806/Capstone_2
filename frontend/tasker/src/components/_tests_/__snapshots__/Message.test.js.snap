// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`smoke and snapshot testing matches snapshot 1`] = `[Function]`;
